numpy/lib/__init__.py

```diff
@@ -11,51 +11,83 @@
 useful to have in the main name-space.
 
 """
-import math
-
-from numpy.version import version as __version__
-
-# Public submodules
-# Note: recfunctions and (maybe) format are public too, but not imported
-from . import mixins
-from . import scimath as emath
-
-# Private submodules
-from .type_check import *
-from .index_tricks import *
-from .function_base import *
-from .nanfunctions import *
-from .shape_base import *
-from .stride_tricks import *
-from .twodim_base import *
-from .ufunclike import *
-from .histograms import *
-
-from .polynomial import *
-from .utils import *
-from .arraysetops import *
-from .npyio import *
+# NOTE: This file is the _public_ interface for additional submodules
+# housed in `numpy.lib`. The private import mechanism lives in the
+# `_lib_importer.py` file, which defines `__all__` to include all names to
+# be added to the main namespace!
+
+from . import scimath as emath  # import explicitly to rename
+
 from .arrayterator import Arrayterator
-from .arraypad import *
-from ._version import *
 from numpy.core._multiarray_umath import tracemalloc_domain
-
-__all__ = ['emath', 'math', 'tracemalloc_domain', 'Arrayterator']
-__all__ += type_check.__all__
-__all__ += index_tricks.__all__
-__all__ += function_base.__all__
-__all__ += shape_base.__all__
-__all__ += stride_tricks.__all__
-__all__ += twodim_base.__all__
-__all__ += ufunclike.__all__
-__all__ += arraypad.__all__
-__all__ += polynomial.__all__
-__all__ += utils.__all__
-__all__ += arraysetops.__all__
-__all__ += npyio.__all__
-__all__ += nanfunctions.__all__
-__all__ += histograms.__all__
+from ._version import NumpyVersion
 
 from numpy._pytesttester import PytestTester
 test = PytestTester(__name__)
 del PytestTester
+
+from . import _lib_importer as _hidden_lib_namespace
+
+
+# Ensure that __all__ only contains those things that we want it to contain.
+# Because this is an __init__, we would otherwise include all submodules.
+#
+# NOTE: The symbols not included in `__all__` will be hidden, but do not go
+#       through `__getattr__` (see `__dir__`).
+__all__ = {'emath', 'Arrayterator', 'tracemalloc_domain', 'NumpyVersion'}
+
+__lib_submodules_ = {
```

Comment:
Out of curiosity, is there a reason why this isn't just a normal single-underscored variable (_e.g._ `_lib_submodules`)?

Reply:
No, I guess I had started with dunders, and then thought that dunders are sometimes special, so no dunder might be better, I suppose it looked very private, like in classes with name mangling :).

I have to think about the approach a bit more.  I am not certain that actually renaming files helps much, this achieves almost the same thing.

@BvB93 yeah, if we do this, the `pyi` file should just be removed, only leaving the things that current are there also. If that doesn't include modules, that only leaves `__all__` and 3 others symbols or so, I guess. I am almost surprised we have a a `pyi` at all...

Reply:
For the `pyi` file: there have been issues in the past with certain circular imports crashing mypy (https://github.com/numpy/numpy/issues/17316), issues that were alleviated by adding the likes `numpy/core/__init__.pyi` file. Because of this I'm a bit reluctant to remove the `pyi` file here in its entirety, as I'm afraid at might break something.